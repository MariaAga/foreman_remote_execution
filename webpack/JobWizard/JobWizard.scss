.job-wizard {
  .pf-c-wizard__main {
    z-index: calc(
      var(--pf-c-wizard__footer--ZIndex) + 1
    ); // So the select box can be shown above the wizard footer
  }

  .pf-c-wizard__main-body {
    max-width: 500px;
    .advanced-fields-title {
      margin-bottom: 10px;
    }
  }
}
