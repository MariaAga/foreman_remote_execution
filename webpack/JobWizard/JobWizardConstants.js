import { foremanUrl } from 'foremanReact/common/helpers';

export const JOB_TEMPLATES = 'JOB_TEMPLATES';
export const JOB_CATEGORIES = 'JOB_CATEGORIES';
export const JOB_TEMPLATE = 'JOB_TEMPLATE';
export const templatesUrl = foremanUrl('/api/v2/job_templates');
