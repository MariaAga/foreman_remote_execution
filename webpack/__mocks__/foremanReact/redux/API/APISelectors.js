export const selectAPIStatus = () => 'RESOLVED';
export const selectAPIResponse = state => state;
export const selectAPIErrorMessage = state => state.error;
