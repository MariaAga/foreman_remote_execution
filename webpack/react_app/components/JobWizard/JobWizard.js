import React, { useState } from 'react';
import { Wizard } from '@patternfly/react-core';
import { translate as __ } from 'foremanReact/common/I18n';
import CategoryAndTemplate from './steps/CategoryAndTemplate/';
import AdvancedFields from './steps/AdvancedFields/';

export const JobWizard = () => {
  const [jobTemplate, setJobTemplate] = useState(null);
  const [category, setCategory] = useState('');
  const [advancedValue, setAdvancedValue] = useState({
    templateValues: {},
    effectiveUserValue: '',
    timeoutToKill: '',
    password: '',
    keyPassphrase: '',
    effectiveUserPassword: '',
    concurrencyLevel: '',
    timeSpan: '',
  });
  const steps = [
    {
      name: __('Category and template'),
      component: (
        <CategoryAndTemplate
          jobTemplate={jobTemplate}
          setJobTemplate={setJobTemplate}
          category={category}
          setCategory={setCategory}
        />
      ),
    },
    {
      name: __('Target hosts'),
      component: <p>TargetHosts </p>,
      canJumpTo: !!jobTemplate,
    },
    {
      name: __('Advanced fields'),
      component: (
        <AdvancedFields
          // jobTemplateID={jobTemplate}
          jobTemplateID={197}
          advancedValue={advancedValue}
          setAdvancedValue={setAdvancedValue}
        />
      ),
      // canJumpTo: !!jobTemplate,
    },
    {
      name: __('Schedule'),
      component: <p>Schedule</p>,
      canJumpTo: !!jobTemplate,
    },
    {
      name: __('Review details'),
      component: <p>ReviewDetails</p>,
      nextButtonText: 'Run',
      canJumpTo: !!jobTemplate,
    },
  ];
  const title = __('Run Job');
  return <Wizard navAriaLabel={`${title} steps`} steps={steps} height="70vh" />;
};

export default JobWizard;
