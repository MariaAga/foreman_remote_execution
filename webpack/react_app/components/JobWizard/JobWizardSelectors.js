import {
  selectAPIResponse,
  selectAPIStatus,
} from 'foremanReact/redux/API/APISelectors';

import { JOB_TEMPLATES, JOB_CATEGORIES, JOB_TEMPLATE } from './JobWizardConsts';

export const selectJobTemplatesStatus = state =>
  selectAPIStatus(state, JOB_TEMPLATES);

export const selectJobTemplates = state =>
  selectAPIResponse(state, JOB_TEMPLATES)?.results?.filter(
    template => !template.snippet
  ) || [];

export const selectJobCategories = state =>
  selectAPIResponse(state, JOB_CATEGORIES).job_categories || [];

export const selectJobCategoriesStatus = state =>
  selectAPIStatus(state, JOB_CATEGORIES);

export const selectJobTemplate = state =>
  selectAPIResponse(state, JOB_TEMPLATE) || {};
