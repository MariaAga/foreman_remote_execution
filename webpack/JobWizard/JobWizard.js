import React, { useState, useEffect, useCallback } from 'react';
import { useDispatch } from 'react-redux';
import { Wizard } from '@patternfly/react-core';
import { get } from 'foremanReact/redux/API';
import { translate as __ } from 'foremanReact/common/I18n';
import history from 'foremanReact/history';
import CategoryAndTemplate from './steps/CategoryAndTemplate/';
import { AdvancedFields } from './steps/AdvancedFields/AdvancedFields';
import { JOB_TEMPLATE } from './JobWizardConstants';
import HostsAndInputs from './steps/HostsAndInputs/';
import './JobWizard.scss';

export const JobWizard = () => {
  const [jobTemplateID, setJobTemplateID] = useState(null);
  const [category, setCategory] = useState('');
  const [advancedValue, setAdvancedValue] = useState({});
  const [templateValues, setTemplateValues] = useState({}); // TODO use templateValues in advanced fields - description
  const [selectedHosts, setSelectedHosts] = useState(['host1', 'host2']);
  const dispatch = useDispatch();

  const setDefaults = useCallback(
    response => {
      const responseJob = response.data;
      const defaultAdvancedTemplateValues = {};
      const defaultTemplateValues = {};
      const inputs = responseJob.template_inputs_with_foreign;
      if (inputs) {
        inputs.forEach(input => {
          if (input.advanced)
            defaultAdvancedTemplateValues[input.name] = input?.default || '';
          else defaultTemplateValues[input.name] = input?.default || '';
        });
      }
      setTemplateValues(defaultTemplateValues);
      setAdvancedValue({
        ...advancedValue,
        effectiveUserValue: responseJob.effective_user?.value || '',
        timeoutToKill: responseJob.job_template?.executionTimeoutInterval || '',
        templateValues,
      });
    },
    [advancedValue, templateValues]
  );
  useEffect(() => {
    if (jobTemplateID) {
      dispatch(
        get({
          key: JOB_TEMPLATE,
          url: `/ui_job_wizard/template/${jobTemplateID}`,
          handleSuccess: setDefaults,
        })
      );
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [jobTemplateID, dispatch]);

  const steps = [
    {
      name: __('Category and template'),
      component: (
        <CategoryAndTemplate
          jobTemplate={jobTemplateID}
          setJobTemplate={setJobTemplateID}
          category={category}
          setCategory={setCategory}
        />
      ),
    },
    {
      name: __('Target hosts and inputs'),
      component: (
        <HostsAndInputs
          templateValues={templateValues}
          setTemplateValues={setTemplateValues}
          selectedHosts={selectedHosts}
          setSelectedHosts={setSelectedHosts}
        />
      ),
      canJumpTo: !!jobTemplateID,
    },
    {
      name: __('Advanced fields'),
      component: (
        <AdvancedFields
          advancedValue={advancedValue}
          setAdvancedValue={newValue => {
            setAdvancedValue(currentAdvancedValue => ({
              ...currentAdvancedValue,
              ...newValue,
            }));
          }}
          jobTemplateID={jobTemplateID}
        />
      ),
      canJumpTo: !!jobTemplateID,
    },
    {
      name: __('Schedule'),
      component: <p>Schedule</p>,
      canJumpTo: !!jobTemplateID,
    },
    {
      name: __('Review details'),
      component: <p>ReviewDetails</p>,
      nextButtonText: 'Run',
      canJumpTo: !!jobTemplateID,
    },
  ];
  const title = __('Run Job');
  return (
    <Wizard
      onClose={() => history.goBack()}
      className="job-wizard"
      navAriaLabel={`${title} steps`}
      steps={steps}
      height="100%"
    />
  );
};

export default JobWizard;
