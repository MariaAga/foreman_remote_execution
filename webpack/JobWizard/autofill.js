import { useEffect } from 'react';
import { useDispatch } from 'react-redux';
import { get } from 'foremanReact/redux/API';
import { HOST_IDS } from './JobWizardConstants';
import './JobWizard.scss';

export const useAutoFill = ({
  fills,
  setSelectedTargets,
  setHostsSearchQuery,
}) => {
  const dispatch = useDispatch();

  useEffect(() => {
    if (Object.keys(fills).length) {
      if (fills['host_ids[]']) {
        dispatch(
          get({
            key: HOST_IDS,
            url: '/api/hosts',
            params: { search: `id = ${fills['host_ids[]'].join(' or id = ')}` },
            handleSuccess: ({ data }) => {
              setSelectedTargets(currentTargets => ({
                ...currentTargets,
                hosts: (data.results || []).map(result => result.name),
              }));
            },
          })
        );
      }
      if (fills.search) {
        setHostsSearchQuery(fills.search);
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
};
