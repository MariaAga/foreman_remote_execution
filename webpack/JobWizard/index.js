import React from 'react';
import PropTypes from 'prop-types';
import URI from 'urijs';
import { Title, Divider } from '@patternfly/react-core';
import { translate as __ } from 'foremanReact/common/I18n';
import PageLayout from 'foremanReact/routes/common/PageLayout/PageLayout';
import { JobWizard } from './JobWizard';

const JobWizardPage = ({ location: { search } }) => {
  const fills = URI.parseQuery(search);
  const title = __('Run job');
  const breadcrumbOptions = {
    breadcrumbItems: [
      { caption: __('Jobs'), url: `/jobs` },
      { caption: title },
    ],
  };
  return (
    <PageLayout
      header={title}
      breadcrumbOptions={breadcrumbOptions}
      searchable={false}
    >
      <React.Fragment>
        <Title headingLevel="h2" size="2xl">
          {title}
        </Title>
        <Divider component="div" />
        <JobWizard fills={fills} />
      </React.Fragment>
    </PageLayout>
  );
};

JobWizardPage.propTypes = {
  location: PropTypes.shape({
    search: PropTypes.string,
  }).isRequired,
};

export default JobWizardPage;
