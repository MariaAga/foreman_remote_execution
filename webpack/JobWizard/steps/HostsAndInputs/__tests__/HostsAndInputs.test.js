import React from 'react';
import { Provider } from 'react-redux';
import { fireEvent, screen, render, act } from '@testing-library/react';
import * as api from 'foremanReact/redux/API';
import { JobWizard } from '../../../JobWizard';
import JobWizardPage from '../../../';
import * as selectors from '../../../JobWizardSelectors';
import { testSetup, mockApi } from '../../../__tests__/fixtures';

const store = testSetup(selectors, api);
mockApi(api);
const lodash = require('lodash');

lodash.debounce = fn => fn;

describe('Hosts', () => {
  it('Host selection chips removal and keep state between steps', async () => {
    render(
      <Provider store={store}>
        <JobWizard />
      </Provider>
    );
    await act(async () => {
      fireEvent.click(screen.getByText('Target hosts and inputs'));
    });
    const select = () => screen.getByRole('button', { name: 'Options menu' });
    fireEvent.click(select());
    await act(async () => {
      fireEvent.click(screen.getByText('host1'));
      fireEvent.click(screen.getByText('host2'));
    });
    fireEvent.click(
      screen.getByText('Hosts', { selector: '.pf-c-select__toggle-text' })
    );
    await act(async () => {
      fireEvent.click(screen.getByText('Host groups'));
    });
    fireEvent.click(select());
    await act(async () => {
      fireEvent.click(screen.getByText('host_group1'));
    });
    fireEvent.click(select());
    expect(screen.queryAllByText('host1')).toHaveLength(1);
    expect(screen.queryAllByText('host2')).toHaveLength(1);
    expect(screen.queryAllByText('host3')).toHaveLength(0);
    const chip1 = screen.getByRole('button', { name: 'Close host1 host1' });
    await act(async () => {
      fireEvent.click(chip1);
    });
    expect(screen.queryAllByText('host1')).toHaveLength(0);
    expect(screen.queryAllByText('host3')).toHaveLength(0);
    expect(screen.queryAllByText('host2')).toHaveLength(1);
    expect(screen.queryAllByText('host_group1')).toHaveLength(1);

    fireEvent.click(
      screen.getByText('Host groups', { selector: '.pf-c-select__toggle-text' })
    );

    await act(async () => {
      fireEvent.click(screen.getByText('Category and Template'));
    });
    await act(async () => {
      fireEvent.click(screen.getByText('Target hosts and inputs'));
    });
    expect(screen.queryAllByText('host2')).toHaveLength(1);
    expect(screen.queryAllByText('host_group1')).toHaveLength(1);
  });
  it('Host fill list from url', async () => {
    render(
      <Provider store={store}>
        <JobWizard fills={{ 'host_ids[]': ['host1', 'host3'] }} />
      </Provider>
    );
    await act(async () => {
      fireEvent.click(screen.getByText('Target hosts and inputs'));
    });
    api.get.mock.calls.forEach(call => {
      if (call[0].key === 'HOST_IDS') {
        expect(call[0].params).toEqual({ search: 'id = host1 or id = host3' });
      }
    });

    expect(screen.queryAllByText('host1')).toHaveLength(1);
    expect(screen.queryAllByText('host2')).toHaveLength(0);
    expect(screen.queryAllByText('host3')).toHaveLength(1);
  });
  it('Host fill search from url', async () => {
    render(
      <Provider store={store}>
        <JobWizard fills={{ search: 'os = gnome' }} />
      </Provider>
    );
    await act(async () => {
      fireEvent.click(screen.getByText('Target hosts and inputs'));
    });
    expect(screen.queryAllByText('os = gnome')).toHaveLength(1);
  });

  it('input fill from url', async () => {
    const inputText = 'test text';
    render(
      <Provider store={store}>
        <JobWizardPage
          location={{
            search: `feature=test_feature&inputs[plain hidden]=${inputText}`,
          }}
        />
      </Provider>
    );
    api.get.mock.calls.forEach(call => {
      if (call[0].key === 'REX_FEATURE') {
        expect(call[0].url).toEqual(
          '/api/remote_execution_features/test_feature'
        );
      }
    });
    await act(async () => {
      fireEvent.click(screen.getByText('Target hosts and inputs'));
    });
    const textField = screen.getByLabelText('plain hidden', {
      selector: 'textarea',
    });
    expect(textField.value).toBe(inputText);
  });
});
