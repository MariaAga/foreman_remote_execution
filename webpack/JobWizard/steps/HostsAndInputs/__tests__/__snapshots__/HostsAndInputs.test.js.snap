// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HostsAndInputs rendering 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <HostsAndInputs
    selectedHosts={
      Array [
        "host1",
      ]
    }
    setSelectedHosts={[MockFunction]}
    setTemplateValues={[MockFunction]}
    templateValues={Object {}}
  >
    <Title
      className="wizard-title"
      headingLevel="h2"
    >
      <h2
        className="pf-c-title pf-m-xl wizard-title"
      >
        Target hosts and inputs
      </h2>
    </Title>
    <Form>
      <form
        className="pf-c-form"
        noValidate={true}
      >
        <mockConstructor
          fieldId="host_selection"
        >
          <div
            props={
              Object {
                "children": Array [
                  <SelectField
                    fieldId="host_methods"
                    label={null}
                    options={
                      Array [
                        "Hosts",
                        "Host collection",
                        "Host group",
                        "Search query",
                      ]
                    }
                    setValue={[Function]}
                    value="Hosts"
                  />,
                  <SelectedChips
                    selected={
                      Array [
                        "host1",
                      ]
                    }
                    setSelected={[MockFunction]}
                  />,
                ],
                "fieldId": "host_selection",
              }
            }
          />
        </mockConstructor>
        <span>
          Apply to
           
          <Button
            isInline={true}
            variant="link"
          >
            <button
              aria-disabled={false}
              aria-label={null}
              className="pf-c-button pf-m-link pf-m-inline"
              data-ouia-component-id="OUIA-Generated-Button-link-1"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe={true}
              disabled={false}
              role={null}
              type="button"
            >
              1
               
              hosts
            </button>
          </Button>
        </span>
        <TemplateInputs
          inputs={Array []}
          setValue={[MockFunction]}
          value={Object {}}
        >
          <p
            className="gray-text"
          >
            There are no available input fields for the selected template.
          </p>
        </TemplateInputs>
      </form>
    </Form>
  </HostsAndInputs>
</Provider>
`;
